src/string/brz.c

```diff
@@ -0,0 +1,696 @@
+#define derive_glob(g, chr) (g)->derive(g,chr)
+
+static int matches(glob const *self, char const *string);
+
+static int is_true(glob const *self);
+static int is_false(glob const *self);
+static int nullable_char(glob const* self);
+static int nullable_or(glob const* self);
+static int nullable_seq(glob const* self);
+static int nullable_first(glob const* self);
+static int nullable_not(glob const* self);
+
+static glob *derive_empty(glob const *self, char const chr);
+static glob *derive_blank(glob const *self, char const chr);
+static glob *derive_char(glob const *self, char const chr);
+static glob *derive_cset(glob const *self, char const chr);
+static glob *derive_or(glob const *self, char const chr);
+static glob *derive_seq(glob const *self, char const chr);
+static glob *derive_plus(glob const *self, char const chr);
+static glob *derive_at(glob const *self, char const chr);
+static glob *derive_not(glob const *self, char const chr);
+static glob *derive_and(glob const *self, char const chr);
+
+static glob *copy_zero(glob const *cpy);
+static glob *copy_one(glob const *cpy);
+static glob *copy_two(glob const *cpy);
+
+#ifdef _DEBUG_BRZ_
+#define print_glob_v(g) do{printf(#g" "); (g)->print(g),puts("");}while(0)
+#define print_glob(g) do{(g)->print(g);}while(0)
+void print_empty(glob const *self);
+void print_blank(glob const *self);
+void print_char(glob const *self);
+void print_cset(glob const *self);
+void print_or(glob const *self);
+void print_and(glob const *self);
+void print_star(glob const *self);
+void print_plus(glob const *self);
+void print_at(glob const *self);
+void print_opt(glob const *self);
+void print_seq(glob const *self);
+void print_not(glob const *self);
+#endif
+
+static glob *new_glob() {
+    glob* tmp = malloc(sizeof(glob));
+    if (tmp) {
+        tmp->derive = NULL;
+        tmp->nullable = NULL;
+        tmp->first = NULL;
+        tmp->second = NULL;
+        tmp->matches = matches;
+        tmp->type = 0;
+    }
+    else {
+        fprintf(stderr, "Could not allocate glob object.\n");
+        exit(1);
+    }
+    return tmp;
+}
+
+void free_glob(glob *tmp) {
+    glob* first = (glob*)tmp->first;
+    glob* second = (glob*)tmp->second;
+    if (tmp->type != CHAR && tmp->type != CHARSET && first) {
+        free_glob(first);
+    } else if (tmp->type == CHARSET) {
+        free(tmp->cset);
+    }
+    if (second) {
+        free_glob(second);
+    }
+    free(tmp);
+}
+
+
+int matches(glob const *self, char const *string) {
+    if (strlen(string) == 0) {
+        return self->nullable(self);
+    } else {
+        int ret;
+#ifdef _DEBUG_BRZ_
+        printf("\nstring: ");
+        puts(string);
+#endif
+        glob *tmp = derive_glob(self, *string);
+#ifdef _DEBUG_BRZ_
+        print_glob_v(tmp);
+#endif
+        ret = tmp->matches(tmp,++string);
+        free_glob(tmp);
+        return ret;
+    }
+}
+
+/*
+ * Empty Set
+ * not nullable
+ * no derivative
+ */
+glob *Empty() {
```

Comment:
Also apply snake_case here. I would recommend adding a prefix like `glob_` to defeat special cases like `Char` or `And` (which is a trigraph on gcc iirc)